[9.0][ADD] web_widget_darkroom: Add o2m DarkroomJS widget

lasley opened the pull request:
# Odoo DarkroomJS Widget

This module provides a [DarkroomJS](https://github.com/MattKetmo/darkroomjs) web widget for use with images fields.

This widget will allow you to perform the following actions on images:
- Zoom
- Rotate
- Crop
- Step back in history client-side (before save)
# Usage

To use this module, you need to:
- Install web_widget_darkroom
- Add the to any One2many image relation by using the `darkroom` widget. Options can be passed through to Darkroom using the `options` key::
  
  <field name="image_id" widget="darkroom"
                         options="{'minWidth': 100}" />

The Odoo DarkroomJS widget passes options directly through to Darkroom, which are copied from the source below::

```
  // Default options
  defaults: {
    // Canvas properties (dimension, ratio, color)
    minWidth: null,
    minHeight: null,
    maxWidth: null,
    maxHeight: null,
    ratio: null,
    backgroundColor: '#fff',

    // Plugins options
    plugins: {},

    // Post-initialisation callback
    initialize: function() { /* noop */ }
  },
```
# Known Issues/Roadmap
- Plugins are not able to be added without inheriting, then redefining the widget in the registry due to JS inheritance.
  *\* This is not scalable because there would need to be an explicit dependency chain in order to avoid registry overwrite.

lasley committed:
[ADD] web_widget_darkroom: Add o2m DarkroomJS widget

lasley commented:
WIP because I would like to add some tests. This has been in use for a while though & works great

pedrobaeza commented:
Isn't better to allow to use Darkroom in any image widget? Adding a button like in this module: https://github.com/OCA/web/pull/395

lasley commented:
Hmmm I do like the thought of implicit usage. 

The one issue I am already running into though is that there are a lot of buttons as part of this widget. I'm going to be fixing the plugin framework too, so there's a lot of places this can go. I'm planning a M2M/O2M widget as well.

I could maybe add a button to pop out a modal w/ this widget embedded in it though. Best of worlds type deal maybe?

So many buttons:
![image](https://cloud.githubusercontent.com/assets/7397578/17915687/506b10ae-6961-11e6-8d14-e4430c65b518.png)

pedrobaeza commented:
I prefer that you add a button in the regular image widget that represents an "edit" action, and clicking on it, a modal screen allows to represent the full Darkroom UI, including Save/Discard buttons. For tree views, you can include a server action that allows to be called within a button with the same effect as clicking the edit button.

lasley commented:
I see where you're coming from @pedrobaeza . My project requires an inline widget for some streamlining of some workflows, however I think I can easily work the button modal operation into scope as well - it will increase the versatility by leaps and bounds.

obulkin committed:
[FIX] web_widget_darkroom: Modal, fixes, cleanup
* Fix bugs involving the crop and pan functionality by modifying crop and zoom
plugins and Darkroom widget
* Add Darkroom modal to normal image widget, using darkroom.modal wizard model
to provide backend support for modal view
* Remove res.users view changes introduced for demo purposes (not needed due
to modal functionality)
* Clean up existing code, removing many unnecessary DarkroomJS files

obulkin commented:
I've added the modal functionality, fixed a few bugs, and generally cleaned up the WIP code, removing a number of unnecessary Darkroom files. The module should be ready for review now, @lasley.

simahawk commented on web_widget_darkroom/README.rst:
```diff
@@ -0,0 +1,89 @@
+  image widget. You should see a Darkoom button that can be clicked to open 
+  the image in a Darkroom modal, where it can be edited and the changes can be 
+  saved.
+
```
can we have some screenshots here?

obulkin replied:
Sure thing. I've added some screenshots to the README.

simahawk replied:
:+1:

simahawk reviewed (approved):
except missing screenshot: code LGTM

moylop260 commented:
Could you check if the new beta messages errors of lint are valid to fix them?
https://travis-ci.org/OCA/web/jobs/211978937

obulkin commented:
@moylop260 - I've made a number of changes based on the linter results. Here's my "beta feedback" about the ESLint rules 😄 :

- Currently the [`multiline-ternary`](http://eslint.org/docs/rules/multiline-ternary) rule is set up to enforce multiline ternaries rather than to prevent them. I'm guessing this is the opposite of what you intended, so the value needs to be changed to `"never"`.
- The [`no-mixed-operators`](http://eslint.org/docs/rules/no-mixed-operators) rule is conflicting with the [`no-extra-parens`](http://eslint.org/docs/rules/no-extra-parens) rule. The latter has a setting that should prevent this, but I would suggest dropping the operator rule entirely, since it's really inflexible and even demands parens in cases like `var a = 5 + 10 / 2;`, where the order of operations is sufficient to make the result obvious.
- The [`no-undefined`](http://eslint.org/docs/rules/no-undefined) rule conflicts with [`no-void`](http://eslint.org/docs/rules/no-void). Maybe drop the latter?

pedrobaeza commented:
@moylop260 can you check these comments about the linter?

moylop260 commented:
@obulkin
Thanks for comment about beta eslint.
I will change a few checks for these cases.
Thanks

:+1:

lasley commented:
LGTM, thanks @obulkin
